### ISSUE TITLE
Semantics of sparse array creation functions

### ISSUE THREAD

**Issue by stefanv** [scipy/scipy#18592]
https://github.com/scipy/scipy/pull/18538 outlines a procedure for (eventually) replacing existing sparse matrix creation functions with their array equivalents. One of the advantages of the proposed approach is that we get to re-examine the interface of those functions in an array context. This issue is to track how the new interface should look.

- `def identity(n, dtype='d', format=None)`

  The 2D identity matrix is by far the most common case, and this API already follows `np.identity`. It can be kept as-is, with the option of adding an `ndim` keyword later.

- `def eye(m, n=None, k=0, dtype=float, format=None)`

  This function has no natural analog in 1D or ND, so can be kept as-is. Nowadays, we tend to allow `func(shape, ...)` instead of `func(s0, s1, ...)`, so we could allow `np.eye((3, 4))` and eventually deprecate the `n` keyword.

- `hstack(blocks, format=None, dtype=None)`, `vstack`: keep as-is; may need to fail for higher dimensional inputs, unless bmat can be generalized. Does not even need an `_array` equivalent, since the inputs determine the output class.

- `block_diag(mats, format=None, dtype=None)`: keep as-is, output class determined by input.

- `random(m, n, density=0.01, format='coo', dtype=None, random_state=None, data_rvs=None)` : shape should be condensed into a single argument; `random_state` is used to sample sparsity structure; unless `_rvs` is a standard pattern, can consider a more intuitive naming for `data_rvs`? The docstring can also be improved, both in description and example, where e.g. one would expect an example such as `data_rvs=partial(rng.poisson, 0.8)`.

- `rand`: deprecate

- `bmat(blocks, format=None, dtype=None)`: provide `barray` with same semantics.

The following functions are listed under sparse creation utilities, but are operations on sparse objects:

```
   kron - kronecker product of two sparse matrices
   kronsum - kronecker sum of sparse matrices
   tril - Lower triangular portion of a matrix in sparse format
   triu - Upper triangular portion of a matrix in sparse format
```

I.e., can be kept as is, with output arrays as appropriate.

Once we've agreed on all interface decisions above, the implementation should be straightforward.

/cc @rossbar @perimosocordiae

**Comment by rossbar** [comment]
The general idea is to have an `_array` version of the sparse creation functions that are the most useful. #18538 did this for `diags` by adding `diags_array`.

IMO the next most important creation functions are `eye` and `random`. `eye` is a more general version of `identity` that allows for non-array squares. In addition, `eye` is part of the [array api spec for creation functions](https://data-apis.org/array-api/2022.12/API_specification/creation_functions.html) whereas `identity` is not. I'd also propose to match the signature from the array api spec. Re: `random`, +1 on specifying a `shape` tuple instead of `m` and `n`.

IMO these two are the most clear-cut (I've already started working on `eye`). There are some functions for which it may make sense not to have an `_array` creation function.

**Comment by stefanv** [comment]
Out of the list, which ones should not have array equivalents. And are there any missing from the list?

**Comment by rossbar** [comment]
> Out of the list, which ones should not have array equivalents. And are there any missing from the list?

IMO, `identity`, `rand` and `spdiags` should not have array equivalents. Maybe `bmat` as well as that seems to be a hold over from matlab/np.matrix; though that's worth discussing further.

**Comment by rossbar** [comment]
> Tl;dr: only random needs to be updated to have a new signature.

I think all of them would have new signatures in the sense that we use this opportunity to introduce positional-only and keyword-only args, as well as rename arguments according to standards where applicable (i.e. `eye(m, n=None)` -> `eye(n_rows, n_cols=None)`)

**Comment by rossbar** [comment]
> @tylerjereddy @rossbar We probably don't want this https://github.com/scipy/scipy/pull/18538 in the 1.11 release.

I'd agree, if it's not too much of a hassle to hold it out. I think it makes sense to try to introduce all of the creation functions all in one release.

**Comment by tylerjereddy** [comment]
If you make PRs against the maintenance branch, try to label them clearly as such and ping me. They probably won't get merged for a bit, and for RC1 I'm already pushing the wheels up, so you'll need to wait for RC2 in two weeks.

Also, you may want to proofread the release notes for `sparse`..

But RC1 is imminent..

**Comment by stefanv** [comment]
@perimosocordiae I think @rossbar and I have general agreement on how to proceed here. Any thoughts before we do?

**Comment by perimosocordiae** [comment]
Overall I agree with the comments above. We may still need to introduce *_array versions to enable parameter cleanup, and to handle cases where dense inputs are provided.

Looking forward to meeting more progress here!

**Comment by perimosocordiae** [comment]
Overall I agree with the comments above. We may still need to introduce *_array versions to enable parameter cleanup, and to cleanly handle cases where dense inputs are provided.

Looking forward to making more progress here!
